braid/residual.c

```diff
@@ -43,22 +44,47 @@ _braid_Residual(braid_Core        core,
    braid_Int        ilower   = _braid_GridElt(grids[level], ilower);
    braid_Real      *ta       = _braid_GridElt(grids[level], ta);
 
+   /* this decides whether Delta correction needs to be computed on this level
+    * If step is called from FRestrict, right after restriction to the coarse grid 
+    * (ie calling_function == braid_ASCaller_Residual)
+    * and the current coarse level == delta_defer_lvl,
+    * then we *don't* need to do anything for Delta correction here.
+    */
+   braid_Int delta_correct;
+   delta_correct = _braid_DoDeltaCorrect(core, level, iter) && !( calling_function == braid_ASCaller_Residual && level == _braid_CoreElt(core, delta_defer_lvl) );
+
    braid_BaseVector rstop;
    braid_Int        ii;
 
    ii = index-ilower;
-   _braid_StepStatusInit(ta[ii-1], ta[ii], index-1, tol, iter, level, nrefine, gupper, status);
-   if ( _braid_CoreElt(core, residual) == NULL )
-   {
-      /* By default: r = ustop - \Phi(ustart)*/
+   
+   if ( delta_correct )
+   {  /* Give the user access to the basis vectors through StepStatusGetBasisVec */
+      _braid_StepStatusInit(ta[ii-1], ta[ii], index-1, tol, iter, level, nrefine, gupper, calling_function, r->basis, status);
+
+      /* By default: r = ustop - \Phi(ustart) */
       _braid_GetUInit(core, level, index, r, &rstop);
-      _braid_BaseStep(core, app,  rstop, NULL, r, level, status);
-      _braid_BaseSum(core, app,  1.0, ustop, -1.0, r);
+      _braid_BaseStep(core, app, rstop, NULL, r, level, status);
+      _braid_CoreFcn(core, sum)(app, 1.0, ustop->userVector, -1.0, r->userVector);
+      _braid_BaseSumBasis(core, app, 0., r->basis, -1.0, r->basis);
```

Comment:
@cadaverous-lives  I'm just curious why the constants are 0 and -1.  I guess you have to flip the sign for the basis in the residual?